From 986707b461eb56d75f55581dd1c8e2633f814795 Mon Sep 17 00:00:00 2001
From: Roger Pau Monne <roger.pau@citrix.com>
Date: Tue, 14 Apr 2026 09:38:55 +0200
Subject: xen/cpu: round cpu_khz calculations
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

All arches truncate the cpu_khz without taking into account the less
significant digits.  Instead use DIV_ROUND() when scaling from Hz to kHz
to get as more accurate kHz value.

Signed-off-by: Roger Pau Monné <roger.pau@citrix.com>
Reviewed-by: Jan Beulich <jbeulich@suse.com>

diff --git a/xen/arch/arm/time.c b/xen/arch/arm/time.c
index dec53b5f7d53..5984aaaba1ac 100644
--- a/xen/arch/arm/time.c
+++ b/xen/arch/arm/time.c
@@ -129,7 +129,7 @@ static void __init preinit_dt_xen_time(void)
     res = dt_property_read_u32(timer, "clock-frequency", &rate);
     if ( res )
     {
-        cpu_khz = rate / 1000;
+        cpu_khz = DIV_ROUND(rate, 1000);
         timer_dt_clock_frequency = rate;
     }
 }
@@ -145,7 +145,7 @@ void __init preinit_xen_time(void)
         preinit_acpi_xen_time();
 
     if ( !cpu_khz )
-        cpu_khz = (READ_SYSREG(CNTFRQ_EL0) & CNTFRQ_MASK) / 1000;
+        cpu_khz = DIV_ROUND(READ_SYSREG(CNTFRQ_EL0) & CNTFRQ_MASK, 1000);
 
     res = platform_init_time();
     if ( res )
diff --git a/xen/arch/x86/time.c b/xen/arch/x86/time.c
index 997d2d132b8e..79fe88ee0983 100644
--- a/xen/arch/x86/time.c
+++ b/xen/arch/x86/time.c
@@ -2288,7 +2288,7 @@ void __init early_time_init(void)
     set_time_scale(&t->tsc_scale, tmp);
     t->stamp.local_tsc = boot_tsc_stamp;
 
-    cpu_khz = tmp / 1000;
+    cpu_khz = DIV_ROUND(tmp, 1000);
     printk("Detected %lu.%03lu MHz processor.\n", 
            cpu_khz / 1000, cpu_khz % 1000);
 
